python/ml4ir/applications/ranking/tests/test_aux_metrics_helper.py

```diff
@@ -206,10 +220,73 @@ def test_compute_aux_label_metrics_invalid_click(self):
+                "AuxIntrinsicFailure": 0.133459,
+                "AuxRankMF": 0.0
             }),
             check_less_precise=True)
 
+    def test_compute_rank_match_failure(self):
+        """Test computation of RankMatchFailure metric"""
+        with self.subTest("Click at rank 1 should get RankMF 0."):
+            self.assertEqual(0.,
+                             compute_rank_match_failure(aux_label_values=pd.Series([]),
+                                                        ranks=pd.Series([]),
+                                                        click_rank=1))
+
+        with self.subTest("Only records above clicked record should be used"):
+            self.assertEqual(compute_rank_match_failure(aux_label_values=pd.Series([2, 0, 4, 5, 6]),
+                                                        ranks=pd.Series([1, 2, 3, 4, 5]),
+                                                        click_rank=3),
+                             compute_rank_match_failure(aux_label_values=pd.Series([2, 0, 4]),
+                                                        ranks=pd.Series([1, 2, 3]),
+                                                        click_rank=3))
+
+        with self.subTest("Records are being sorted correctly before computation"):
+            self.assertEqual(compute_rank_match_failure(aux_label_values=pd.Series([2, 3, 0, 5, 6]),
+                                                        ranks=pd.Series([1, 2, 3, 4, 5]),
+                                                        click_rank=5),
+                             compute_rank_match_failure(aux_label_values=pd.Series([5, 6, 0, 3, 2]),
+                                                        ranks=pd.Series([4, 5, 3, 2, 1]),
+                                                        click_rank=5))
+
+        with self.subTest(
+                "If all records above clicked record have aux values greater than 0, then score should be 0."):
+            self.assertEqual(0.,
+                             compute_rank_match_failure(aux_label_values=pd.Series([1, 1.1, 1.2, 0., 0.]),
+                                                        ranks=pd.Series([1, 2, 3, 4, 5]),
+                                                        click_rank=3))
+
+        with self.subTest("Same ordering of aux scores with different values produce same scores"):
+            self.assertEqual(compute_rank_match_failure(aux_label_values=pd.Series([5, 0, 3, 4, 1]),
+                                                        ranks=pd.Series([1, 2, 3, 4, 5]),
+                                                        click_rank=5),
+                             compute_rank_match_failure(aux_label_values=pd.Series([10, 0, 6, 8, 2]),
+                                                        ranks=pd.Series([1, 2, 3, 4, 5]),
+                                                        click_rank=5))
+
+        with self.subTest("Dense method should be used to resolve ties in ranking aux labels"):
```

Comment:
@lastmansleeping 
'Dense method" What does this mean?

Reply:
```
method{‘average’, ‘min’, ‘max’, ‘dense’, ‘ordinal’}, optional
The method used to assign ranks to tied elements. The following methods are available (default is ‘average’):

‘average’: The average of the ranks that would have been assigned to all the tied values is assigned to each value.

‘min’: The minimum of the ranks that would have been assigned to all the tied values is assigned to each value. (This is also referred to as “competition” ranking.)

‘max’: The maximum of the ranks that would have been assigned to all the tied values is assigned to each value.

‘dense’: Like ‘min’, but the rank of the next highest element is assigned the rank immediately after those assigned to the tied elements.

‘ordinal’: All values are given a distinct rank, corresponding to the order that the values occur in a.

```

Link -> https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rankdata.html